src/cow.jl

```diff
@@ -71,117 +93,125 @@ maybe_copy(x::AbstractDict) = begin
 end
 
 # (func_name => mutated_arg_pos)
-const MUTATING_OPS = Dict{Symbol, Int}(
-    :setindex! => 1,
-    :push! => 1, :pushfirst! => 1,
-    :pop! => 1, :popfirst! => 1,
-    :append! => 1,
-    :delete! => 1, :deleteat! => 1,
-    :setdiff! => 1,
+const MUTATING_OPS = Dict{Any, Vector{Int}}(
+    :setindex! => [1],
+    :push! => [1], :pushfirst! => [1],
+    :pop! => [1], :popfirst! => [1],
+    :append! => [1],
+    :delete! => [1], :deleteat! => [1],
+    :setdiff! => [1],
 )
 
 """
     mutating(e)
 
 Predicate if an expression is a mutating operation.
 If true, returns `ture` and the position of the mutated argument;
-else returns `(false, 0)`.
+else returns `(false, [0])`.
 
 Examples:
 
-- `push!(a, v)` -> (true, 1)
-- `print(a)` -> (false, 0)
+- `push!(a, v)` -> (true, [1])
+- `print(a)` -> (false, [0])
 
 """
 function mutating(e)
-    isa(e, Expr) || return (false, 0)
-    e.head == :call || return (false, 0)
-    if isa(e.args[1], GlobalRef) && haskey(MUTATING_OPS, e.args[1].name)
-        return (true, MUTATING_OPS[e.args[1].name])
+    isa(e, Expr) || return (false, [0])
+    e.head == :call || return (false, [0])
+    (args, offset) = if e.args[1] === IRTools.self
+        (e.args[2:end], 1)
+    else
+        (e.args[1:end], 0)
     end
 
-    # IRTools.Statement(:((IRTools.Self())(Main.push!, %2, 1)), Any, 2)
-    if (e.args[1] === IRTools.self) && isa(e.args[2], GlobalRef) &&
-        haskey(MUTATING_OPS, e.args[2].name)
-        return (true, MUTATING_OPS[e.args[2].name] + 1)
-    end
-    return (false, 0)
-end
+    #@show args[1]
 
-function _successors(b::IRTools.Block, accu::Vector{Int64})
-    succs = IRTools.successors(b)
-    ids = map(x -> x.id, succs)
-    issubset(ids, accu) && return
-    for blk in succs
-        push!(accu, blk.id)
-        _successors(blk, accu)
+    if isa(args[1], GlobalRef)
+        if haskey(MUTATING_OPS, args[1])
+            return (true, MUTATING_OPS[args[1]] .+ offset)
+        elseif haskey(MUTATING_OPS, args[1].name)
+            return (true, MUTATING_OPS[args[1].name] .+ offset)
+        end
```

Comment:
```suggestion
    fn = isa(args[1], GlobalRef) ? args[1].name : args[1]
    if haskey(MUTATING_OPS, fn)
        return (true, MUTATING_OPS[fn] .+ offset)
    end
```

Or even have a separate `remove_global` function, if this is needed more than once.

Reply:
Well, the two versions are not the same as what they are doing.
This piece of bad-smelling code is here because I put both symbols and `GlobalRef`s in `MUTATING_OPS`:

```
MUTATING_OPS = Dict(
:push! => [...],
GlobalRef(Mod1, :fund) => [...]
)
```
I wanted to make all the keys `GlobalRef`, but I find the function `Base.push!` is sometimes referred to as `AnotherModuleThanBase.push!`, so I can't tell if it is `Base.push!`, is there an elegant way to deal with this?

Thanks @phipsgabler